### ISSUE TITLE
build warnings / errors

### ISSUE THREAD

**Issue by neumannoskar** [unrealgt/unrealgt#8]
Hello, thank you for putting this plug-in as open source free for everyone. If everything works as expected I will have a good milestone of my master thesis done already!

So I'm using Windows 10 with Unreal Engine 4.26.0 and Visual Studio 2019. I've been trying to build the plug-in as described. During step 3 of your installation guide I get the following warnings when building the visual studio solution:

UnrealBuildTool : warning : Trying to build an enterprise target but the enterprise directory is missing. Falling back on engine components only.

Also I'm not quite familiar with visual studio. Do I select the target for 'Run & Build the solution' here: 

![Unbenannt](https://user-images.githubusercontent.com/57449854/104225263-75755600-5446-11eb-8134-918eee2318cd.JPG) ?

Also this error shows up when I try running the solution after building:

![Unbenannt1](https://user-images.githubusercontent.com/57449854/104225581-de5cce00-5446-11eb-93c2-2bb46d4ebb16.JPG)

And since this failed plug-in installation when I try to open my project in the Unreal Editor simply through the Epic Games Launcher (which was no problem before), I'll get following error:

![Unbenannt](https://user-images.githubusercontent.com/57449854/104226448-1b759000-5448-11eb-8db1-6682945f10c7.JPG)

Thank's for any reply, I really appreciate it.

**Comment by lolleko** [comment]
You should build an `Editor` Target since the plugin is only supported within the editor. To run it as a standalone game some additional setup might be required I have never tried that.

Do you get any C++ errors when Building and Launching the editor via VStudio?

Also I haven't updated for 4.26 yet so chances are the plugin is broken with that version. This is usually very easy to fix, so if you can provide some more information I can start working on a fix.

**Comment by neumannoskar** [comment]
Hey, thank's for the quick answer. So if I chose with DebugGameEditor:

![Unbenannt](https://user-images.githubusercontent.com/57449854/104289293-3cc29480-54b9-11eb-96a6-501c5fc3c632.JPG)

and try to build the solution:

![Unbenannt1](https://user-images.githubusercontent.com/57449854/104289389-5b289000-54b9-11eb-8ba1-82f9e8ddd111.JPG)

I get following output:

1>------ Skipped Build: Project: UE4, Configuration: BuiltWithUnrealBuildTool Win32 ------
1>Project not selected to build for this solution configuration 
2>------ Build started: Project: MyProject, Configuration: DebugGame_Editor x64 ------
2>Creating makefile for MyProjectEditor (no existing makefile)
2>UnrealBuildTool : warning : Trying to build an enterprise target but the enterprise directory is missing. Falling back on engine components only.
2>Parsing headers for MyProjectEditor
2>  Running UnrealHeaderTool "C:\Users\Hallo.DESKTOP-4OVSO14\Documents\Unreal Projects\MyProject\MyProject.uproject" "C:\Users\Hallo.DESKTOP-4OVSO14\Documents\Unreal Projects\MyProject\Intermediate\Build\Win64\MyProjectEditor\DebugGame\MyProjectEditor.uhtmanifest" -LogCmds="loginit warning, logexit warning, logdatabase error" -Unattended -WarningsAsErrors -abslog="C:\Users\Hallo.DESKTOP-4OVSO14\AppData\Local\UnrealBuildTool\Log_UHT.txt" -installed
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Misc/GTRandomObjectPlacer.h(20) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Misc/GTRandomObjectPlacer.h(23) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Misc/GTRandomObjectPlacer.h(26) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Misc/GTRandomObjectPlacer.h(29) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Misc/GTRandomObjectPlacer.h(32) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Misc/GTRandomObjectPlacer.h(35) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Misc/GTRandomObjectPlacer.h(38) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Streamers/GTDataStreamerComponent.h(36) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Streamers/GTFileStreamerComponent.h(31) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Streamers/GTSimpleTCPStreamer.h(36) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:/Program Files/Epic Games/UE_4.26/Engine/Plugins/unrealgt-master/Source/UnrealGT/Public/Streamers/GTSimpleTCPStreamer.h(39) : LogCompile: Error: An explicit Category specifier is required for any property exposed to the editor or Blueprints in an Engine module.
2>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.MakeFile.Targets(46,5): error MSB3073: The command "C:\WINDOWS\System32\chcp.com 65001 >NUL 
2>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.MakeFile.Targets(46,5): error MSB3073:  "C:\Program Files\Epic Games\UE_4.26\Engine\Build\BatchFiles\Build.bat" MyProjectEditor Win64 DebugGame -Project="C:\Users\Hallo.DESKTOP-4OVSO14\Documents\Unreal Projects\MyProject\MyProject.uproject" -WaitMutex -FromMsBuild" exited with code 6.
2>Done building project "MyProject.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 1 skipped ==========



Also following error:
`
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB3073	The command "C:\WINDOWS\System32\chcp.com 65001 >NUL 
 "C:\Program Files\Epic Games\UE_4.26\Engine\Build\BatchFiles\Build.bat" MyProjectEditor Win64 DebugGame -Project="C:\Users\Hallo.DESKTOP-4OVSO14\Documents\Unreal Projects\MyProject\MyProject.uproject" -WaitMutex -FromMsBuild" exited with code 6.	MyProject	C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.MakeFile.Targets	46	
`
Please find a copy of the v160-folder down here.
[v160.zip](https://github.com/unrealgt/unrealgt/files/5800806/v160.zip)

I hope this does provide the necessary since my skills in visual studio are close to none (I'm still not quite sure what you meant with 'build and `Editor` Target'). Thank's for the help!

**Comment by lolleko** [comment]
Okay, this is probably because 4.26 now requires "Category" to be set for UPROPERTIES.

You can fix these issues by adding ` Category = SomeCategoryName` to the UPROPERTY() macro in the reported lines

e.g.

https://github.com/unrealgt/unrealgt/blob/26e882346c7e282ff067907b055e7d2ddbfa5282/Source/UnrealGT/Public/Misc/GTRandomObjectPlacer.h

```
USTRUCT()
struct UNREALGT_API FGTRandomObjectSpawnInfo {
  GENERATED_BODY()

public:
  UPROPERTY(EditAnywhere)
  UStaticMesh *Mesh;

  UPROPERTY(EditAnywhere)
  float Weight;

  UPROPERTY(EditAnywhere)
  float Radius;

  UPROPERTY(EditAnywhere)
  bool bShouldAlignToGround;

  UPROPERTY(EditAnywhere)
  bool bRandomYaw;

  UPROPERTY(EditAnywhere)
  TArray<UMaterialInterface *> RandomMaterials;

  UPROPERTY(EditAnywhere)
  int RandomMaterialSlotID;

  FGTRandomObjectSpawnInfo()
      : Mesh(nullptr), Weight(10.f), Radius(140.f), bShouldAlignToGround(true),
        bRandomYaw(true), RandomMaterials(), RandomMaterialSlotID(0) {}
};
```

=>

```
USTRUCT()
struct UNREALGT_API FGTRandomObjectSpawnInfo {
  GENERATED_BODY()

public:
  UPROPERTY(EditAnywhere, Category = SpawnInfo)
  UStaticMesh *Mesh;

  UPROPERTY(EditAnywhere, Category = SpawnInfo)
  float Weight;

  UPROPERTY(EditAnywhere, Category = SpawnInfo)
  float Radius;

  UPROPERTY(EditAnywhere, Category = SpawnInfo)
  bool bShouldAlignToGround;

  UPROPERTY(EditAnywhere, Category = SpawnInfo)
  bool bRandomYaw;

  UPROPERTY(EditAnywhere, Category = SpawnInfo)
  TArray<UMaterialInterface *> RandomMaterials;

  UPROPERTY(EditAnywhere, Category = SpawnInfo)
  int RandomMaterialSlotID;

  FGTRandomObjectSpawnInfo()
      : Mesh(nullptr), Weight(10.f), Radius(140.f), bShouldAlignToGround(true),
        bRandomYaw(true), RandomMaterials(), RandomMaterialSlotID(0) {}
};
```

I currently don't have the time to fix this. SO I hope you will be able to fix it yourself by adding the category field everywhere it is missing. A PR would be welcome aswell.

> I'm still not quite sure what you meant with 'build an Editor Target'

that just means that you should select a build target from the dropdown in visual studio that includes "Editor", "DebugGame Editor" for example.

**Comment by neumannoskar** [comment]
> A PR would be welcome aswell.

If PR means publication reference, you will definitely be cited and your project and the associated conference paper will be mentioned. 

Unfortunately, your suggestion does not fix the problem. I get the same error. However, since the project is now somewhat in the past, I of course understand that you do not have time to fix this immediately. Nevertheless, I thank you for any help.

**Comment by lolleko** [comment]
> If PR means publication reference, you will definitely be cited and your project and the associated conference paper will be mentioned.

I mean a pull request on GitHub, you uploading the patch that fixes the build errors :D But a citation is obviously also welcome.

> Unfortunately, your suggestion does not fix the problem. I get the same error. However, since the project is now somewhat in the past, I of course understand that you do not have time to fix this immediately. Nevertheless, I thank you for any help.

Have you fixed **every** occurrence of the error? its in multiple files not only `GTRandomObjectPlacer.h` (see error message). This should fix the issue, or at the very least lead to a different/new error.

**Comment by neumannoskar** [comment]
Alright, I see. I now did the changes in all files and the build was successful this time. Though trying to run by pressing this button

![1](https://user-images.githubusercontent.com/57449854/104358096-49bda300-550e-11eb-9361-4a55a3771880.JPG)

Produces this error

![Unbenannt](https://user-images.githubusercontent.com/57449854/104358172-5cd07300-550e-11eb-9817-6b5a7f90b48e.JPG)

So I'm down one missing module compared to before but still not at the end. 

See output log 
[output.txt](https://github.com/unrealgt/unrealgt/files/5804125/output.txt)

I can push as soon as I'm running with version 4.26.

**Comment by lolleko** [comment]
Mhhh that error means that UnrealGt was not built, are you sure there are no more errors when you build (Ctrl+Sift+B)?
also try adding the plugin to our .uproject:


```
{
       ....
	"Modules": [...],
	"Plugins": [
		{
			"Name": "UnrealGT",
			"Enabled": true
		},
	]
}
```

**Comment by neumannoskar** [comment]
Seems like it yeah. See the .uproject and a screen record of the build.

[MyProject.uproject.txt](https://github.com/unrealgt/unrealgt/files/5804883/MyProject.uproject.txt)

https://user-images.githubusercontent.com/57449854/104376440-4f72b300-5525-11eb-8a03-453c6d02132c.mp4

**Comment by lolleko** [comment]
Did you regenerate your project files after adding the plugin? Right click your .uproject file and click "generate visual studio project", before you do that also delete your Intermediate and Binaries folder.

I can try building the plugin myself tomorrow, to check if this is an issue caused by the plugin or by your build setup.

**Comment by neumannoskar** [comment]
That would be great. After deleting the Intermediate and Binaries folder, generating visual studio project and building the solution they is some stuff in the newly created Intermediate and Binaries missing compared to before:

![Binaries Comparison](https://user-images.githubusercontent.com/57449854/104427297-5f20e480-5583-11eb-9f35-662aef510d88.JPG)

![Intermediate Comparison](https://user-images.githubusercontent.com/57449854/104427315-60eaa800-5583-11eb-9f3b-0fbdb04e6399.JPG)
(left is before, right is newly created)

Also I now get an error I've never had before:

![Unbenannt](https://user-images.githubusercontent.com/57449854/104427488-97c0be00-5583-11eb-85f8-f9e6ab5b88d0.JPG)

I'm not sure if it does matter but I installed Unreal Engine Version 4.24.3 meanwhile. Maybe this caused some irritation with the UE4.exe. Would be great if you could give it a try since my knowledge with C++, Visual Studio and the Unreal API is limited. For the case that we get stuck here, I'll just try and build the Plug-in with the older version.

**Comment by lolleko** [comment]
Right click "MyProject" in Visual Studio and "Select as active project". After regenerating VS false chooses UE4 as main project and not your project ("MyProject"). After that you should be able to build again.

Those files are missing because you haven't build yet.

**Comment by neumannoskar** [comment]
The error of not finding the .exe path occurred after the build when I tried running the solution. With your suggestion (couldn't find > "Select as active project" < so I did "Set as Startup Project") I did run again in the missing GT module problem:

**Comment by lolleko** [comment]
I created a fresh 4.26 project and tried to reproduce your issues, here are the exact steps:

1. Create new Blank Unreal C++ project
2. Create new folder “Plugins” inside new Project
3. Clone unrealgt master into Plugins folder
4. Rightclick .uproject => “Generate Visual Studio Project files”
5. Try Build “Development Editor” Target
6. Fixed some build issues (now on master 505cb5fe55ba1c6501a15aa2863bd54c04c194e1)
    1. I did not encounter any missing category errors (like you did, so not sure what is going on there)
    2. But I fixed a bunch of other issues which might be the reason your build failed
7. Build again and Launched
8. Everything seems to work

Note: I did not edit my .uproject file and did not enable the plugin. Just putting it into the plugins dir and building it together with the project seems to be enough

**Comment by neumannoskar** [comment]
Done, good job!

![Unbenannt](https://user-images.githubusercontent.com/57449854/104463275-dbcab780-55b1-11eb-9215-60b15da4ce76.JPG)

I guess the problem with 'MyProject' from before was that I used the 'Architecture, Engineering and Construction' as New Project Category. My plan is to generate synthetic dataset from .ifc - models (buildings or construction sites). Are they huge differences in the categories, could I be running in some problems with importing models from .ifc-files using the 'Games' category?
